lib/ecto/changeset.ex

```diff
@@ -526,8 +526,10 @@ defmodule Ecto.Changeset do
         {Map.put(changes, key, value), errors, valid?}
       :missing ->
         {changes, errors, valid?}
-      :invalid ->
-        {changes, [{key, {"is invalid", [type: type, validation: :cast]}} | errors], false}
+      {:invalid, custom_errors} ->
+        {message, custom_errors} = Keyword.pop(custom_errors, :message, "is invalid")
+        new_errors = [type: type] ++ Keyword.put_new(custom_errors, :validation, :cast)
```

Comment:
If the user provides the `:type` key, the resulting error will contain two `:type` keys, which might lead to some side effects similar to overriding the `:type` key. For example:

```elixir
iex(1)> errors = [type: MyType] ++ [type: MyErrorType, validation: :my_validation]
[type: MyType, type: MyErrorType, validation: :my_validation]
iex(2)> Enum.into(errors, %{})
%{type: MyErrorType, validation: :my_validation}
```